test-data/unit/fine-grained.test

```diff
@@ -1743,6 +1743,292 @@ def f(x=1): pass
+[file submod.py.3]
+class M(type):
+    y = None  # type: str
+[file submod.py.4]
+class M(type):
+    x = None  # type: int
+[out]
+==
+a.py:5: error: Incompatible types in assignment (expression has type "int", variable has type "str")
+==
+a.py:5: error: "Type[C]" has no attribute "x"
+==
+
+[case testMetaclassAttributesDirect_python2]
+# flags: --py2
+import a
+[file a.py]
+from mod import C
+def f():
+    # type: () -> None
+    C.x = int()
+[file mod.py]
+import submod
+class C:
+    __metaclass__ = submod.M
+[file submod.py]
+class M(type):
+    x = None  # type: int
+[file submod.py.2]
+class M(type):
+    x = None  # type: str
+[file submod.py.3]
+class M(type):
+    y = None  # type: str
+[file submod.py.4]
+class M(type):
+    x = None  # type: int
+[out]
+==
+a.py:4: error: Incompatible types in assignment (expression has type "int", variable has type "str")
+==
+a.py:4: error: "Type[C]" has no attribute "x"
+==
+
+[case testMetaclassOperators_python2]
+# flags: --py2
+import a
+[file a.py]
+from mod import C
+from typing import Type
+def f(arg):
+    # type: (Type[C]) -> None
+    arg + arg
+[file mod.py]
+import submod
+class C:
+    __metaclass__ = submod.M
+[file submod.py]
+class M(type):
+    def __add__(self, other):
+        # type: (M) -> M
+        pass
+[file submod.py.2]
+class M(type):
+    def __add__(self, other):
+        # type: (int) -> M
+        pass
+[out]
+==
+a.py:5: error: Unsupported operand types for + ("Type[C]" and "Type[C]")
+
+[case testMetaclassOperatorsDirect_python2]
+# flags: --py2
+import a
+[file a.py]
+from mod import C
+def f():
+    # type: () -> None
+    C + C
+[file mod.py]
+import submod
+class C:
+    __metaclass__ = submod.M
+[file submod.py]
+class M(type):
+    def __add__(self, other):
+        # type: (int) -> M
+        pass
+[file submod.py.2]
+class M(type):
+    def __add__(self, other):
+        # type: (M) -> M
+        pass
+[out]
+a.py:4: error: Unsupported operand types for + ("Type[C]" and "Type[C]")
+==
+
+[case testFineMetaclassUpdate]
+import a
+[file a.py]
+from b import B
+B.x
+
+[file b.py]
+import c
+class B(metaclass=c.M): pass
+
+[file c.py]
+class M(type):
+    x: int
+
+[file c.py.2]
+class M(type):
+    y: int
+[out]
+==
+a.py:2: error: "Type[B]" has no attribute "x"
+
+[case testFineMetaclassUpdate2]
+import a
+[file a.py]
+from c import M
+import b
+def f(arg: M) -> None:
+    pass
+
+f(b.B)
+[file b.py]
+import c
+class B: pass
+
+[file b.py.2]
+import c
+class B(metaclass=c.M): pass
+
+[file c.py]
+class M(type):
+    pass
+[out]
+a.py:6: error: Argument 1 to "f" has incompatible type "Type[B]"; expected "M"
+==
+
+[case testFineMetaclassRecalculation]
+import a
+[file a.py]
+from b import B
+class M2(type): pass
+class D(B, metaclass=M2): pass
+[file b.py]
+import c
+class B: pass
+
+[file b.py.2]
+import c
+class B(metaclass=c.M): pass
+
+[file c.py]
+class M(type):
+    pass
+[out]
+==
+a.py:3: error: Inconsistent metaclass structure for 'D'
```

Comment:
Ideas for additional tests cases:

* Removing metaclass from a class.
* Change metaclass to another one.
* Initially metaclass definition doesn't point to a class (undefined or not a class). In update point metaclass to a suitable class.
* Update makes metaclass point to a non-class.